### ISSUE TITLE
Startup is 44x slower when using MMTk

### ISSUE THREAD

**Issue by chrisseaton** [mmtk/mmtk-ruby#13]
What causes this huge difference in startup time when MMTk is enabled? 0.01 to 0.6s, or 44x. That's slow enough to be very noticeable interactively, and it also makes our simplest test suite very slow as it runs a new Ruby for each test.

```
% time ruby/build/bin/ruby --version
ruby 3.2.0dev (2022-08-31T08:01:00Z merge-upstream 04cc11c303) [x86_64-linux]
ruby/build/bin/ruby --version  0.01s user 0.01s system 98% cpu 0.013 total

% time ruby/build/bin/ruby --mmtk --version
ruby 3.2.0dev (2022-08-31T08:01:00Z merge-upstream 04cc11c303) +MMTk(MarkSweep) [x86_64-linux]
ruby/build/bin/ruby --mmtk --version  0.11s user 0.44s system 97% cpu 0.568 total
```

**Comment by steveblackburn** [comment]
One possibility is a bug that Zixian found a few weeks back.   Metadata was being allocated for all of _virtual memory_ (!), which meant many 100's of MB of metadata allocated and initialized at boot time.

**Comment by qinsoon** [comment]
> One possibility is a bug that Zixian found a few weeks back. Metadata was being allocated for all of _virtual memory_ (!), which meant many 100's of MB of metadata allocated and initialized at boot time.

Yeah, that should be the reason. There is [a pull request](https://github.com/mmtk/mmtk-core/pull/632) to fix it. I will try get it merged soon.

**Comment by chrisseaton** [comment]
Our cloud development instances have 125 GB of RAM, so the default heap size is very large heap, which probably exacerbates that bug.

**Comment by eightbitraptor** [comment]
This appears to have addressed some of the slowdown but we're still seeing 0.391s with `--mmtk` vs 0.015s without

```
% time ./ruby -v
ruby 3.3.0dev (2023-01-10T12:30:22Z mmtk 2e90a2b260) [x86_64-linux]
./ruby -v  0.01s user 0.01s system 98% cpu 0.015 total
% time ./ruby -v --mmtk
ruby 3.3.0dev (2023-01-10T12:30:22Z mmtk 2e90a2b260) +MMTk(MarkSweep) [x86_64-linux]
./ruby -v --mmtk  0.08s user 0.29s system 94% cpu 0.391 total
```

**Comment by wks** [comment]
I found one source of the start-up overhead, and I have submitted a PR for that. https://github.com/mmtk/mmtk-core/pull/737

The other source of slowness is related to the way we call `mmap`. During start-up, the `FragmentedMapper` tries to reserve chunks of memory using `mmap` with `PROT_NONE`.  However, it calls one `mmap` for every 4MiB chunk. Although each call costs only 1.1us, it makes a total of 57769 such calls during the creating of MMTk.  That sums up to a very considerable amount.  The obvious solution is to coalesce `mmap` calls for adjacent chunks.
